### ISSUE TITLE
Slow UI (switching notes, opening search, etc.)

### ISSUE THREAD

**Issue by Klemet** [laurent22/joplin#5563]
Not sure if this is a bug, or a limitation of electron-based apps.

I've been using Joplin for more than a year now. I've always noticed the UI being pretty unresponsive or slow when compared to other apps that I use.

- Switching from one note to another takes at least a good second (from the moment the click is done).
   - This is cumulative when switching multiple times: I have to wait for the switches to be done.
   - This is also the case when switching notes with the "Note Tabs" plugin.
- Switching notebooks, or folding/unfolding them also have a noticeable lag between the click and the response
- Opening the search panel (ctrl+p) often takes a good second too.
- Adding tags is slow (see #4251)
- There often are slowdowns when writing a note - the characters imputed on the keyboard take some time to finally display in the note.

Am I the only one with this issue ? Is it normal ? Or am I too demanding (which I would understand) ?

## Environment

Joplin version: 2.4.9
Platform: Windows
OS specifics: Windows 10 

## Steps to reproduce

1. Use Joplin
2. Notice moments when UI is more un-responsive than usual for app 

## Describe what you expected to happen

- A faster switch between notes, with less noticeable delay
- An overall more responsive UI

**Comment by JackGruber** [comment]
I'll see if I can test this by deleting all tags in a copy of my profile and then making a comparison.

**Comment by Klemet** [comment]
> 
> 
> Any plugins? And any error in the console? https://joplinapp.org/debugging
> 
> Also please provide the number of items, as it appears in Help => Synchronisation Status

Sorry, I should have added more infos !

- 15 plugins enabled
- No error in the console, except this repeated warning:

> DevTools failed to load SourceMap: Could not load content for file:///C:/Program%20Files/Joplin/resources/app.asar/index.js.map:

And this other warning:

> Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.

- 5447 total items.

**I think I found out the issue**, though ! To be honest, it did not even cross my mind that it could come from one of the plugins that I've been using for a long time.
**By disabling all plugins, and especially Link Graph UI**, the UI becomes much more responsive. Switching notes and notebooks or opening the search bar is basically as fast as I would hope it to be, The tag problem is still here, but is truly tolerable when compared to the slow down that I was encountering in moving between notes.

> In which editor?

I was encountering slowdowns in the markdown editor; but maybe this is related to plugins, though. It's hard to reproduce, as I haven't pin-pointed the moments when it happens.

* * *

Anyway; Laurent, I'm sorry if I made you lose some time 🙏. I think that the issue might be closed now, for what concerns me. The moral of the story would be to expect trade-offs between the quantity of plugins and some aspects of Joplin's performance.

**Comment by laurent22** [comment]
@Klemet, just to be sure could you try with and without just Link Graph UI, to check that it's indeed the plugin causing this?

@JackGruber, interesting, it looks like there's something very inefficient about the way tags are handled. I wonder if it's related to the note counter, which maybe updates any time something is changed.

**Comment by Klemet** [comment]
I tried to use the "Performance" menu of the developer tools to see if I could get some data on what happens with or without the Link Graph UI plugin.

I took two recording of 12 seconds during which I switched notes, giving them time to load.

Here is the result : 

![image](https://user-images.githubusercontent.com/41070328/137230554-48231c20-fe5d-4a32-9e68-a6ec581cabe8.png)

With the Link Graph UI plugin, there seems to be:

- Lots of tiny spikes in the number of Listeners (blue curve) (Or is it JS Heap ? Not sure)
- Less idle time (but this might be me not timing the note switch right between the two recordings)

I imagine that there are more ways to check out differences in performances, in functions called, etc.; sadly, I don't have the time to delve further on it tonight.
